### PR TITLE
rust: add simple transform and produce strategy example

### PR DESCRIPTION
**Pull request by dbanda**
Adds the produce strategy and a simple transform -> produce example. 

The produce strategy is an arroyo strategy that produces messages on a given topic.

### CODE REVIEW

File: rust_snuba/rust_arroyo/src/processing/strategies/produce.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    pub producer: KafkaProducer,
+    pub next_step: Box<dyn ProcessingStrategy<TPayload>>,
+    queue: VecDeque<Message<TPayload>>,
+    topic: TopicOrPartition,
+    closed: bool,
+    max_queue_size: usize,
+}
+
+// TODO: make the queue store futures
+// TODO: make this generic
```
Comment (lynnagara) [inline rust_snuba/rust_arroyo/src/processing/strategies/produce.rs:23]: Does it actually need to be generic? i.e. Any reason to take anything other than a KafkaPayload here?

File: rust_snuba/rust_arroyo/src/processing/strategies/produce.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        return None
+    }
+
+    fn submit(&mut self, message: Message<KafkaPayload>) -> Result<(), MessageRejected> {
+        if self.queue.len() >= self.max_queue_size {
+            return Err(MessageRejected)
+        }
+        print!("Producing message: {:?}", &message.payload.payload);
+        self.producer.produce(&self.topic, &message.payload);
+        self.producer.flush();
```
Comment (lynnagara) [inline rust_snuba/rust_arroyo/src/processing/strategies/produce.rs:55]: `submit` is supposed to be a non blocking method, here we're waiting indefinitely for the message to get flushed

File: rust_snuba/rust_arroyo/src/processing/strategies/produce.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        self.queue.push_back(message);
+        return Ok(());
+    }
+
+    fn close(&mut self) {
+        self.closed = true;
```
Comment (lynnagara) [inline rust_snuba/rust_arroyo/src/processing/strategies/produce.rs:61]: `self.closed` isn't being used anywhere

**Review by lynnagara — COMMENTED** [review]
Shall we decide how we want to deal with the asynchronous produce step?

A couple of option I can see:

1 - Have this strategy create the producer inside here instead of getting it passed in from outside by the user of this strategy. This would allow us to pass it the delivery callbacks we need so that we can get notified when messages have been successfully produced.

2 - Change producer.produce() to return a future, and avoid those callbacks.

What do you think?

**Review by untitaker — APPROVED** [review]
just gonna approve so we can merge it for now, i havent' read it in detail

**Comment by dbanda** [comment]
> Shall we decide how we want to deal with the asynchronous produce step?
> 
> A couple of option I can see:
> 
> 1 - Have this strategy create the producer inside here instead of getting it passed in from outside by the user of this strategy. This would allow us to pass it the delivery callbacks we need so that we can get notified when messages have been successfully produced.
> 
> 2 - Change producer.produce() to return a future, and avoid those callbacks.
> 
> What do you think?

My preference would be for 2. Passing callbacks in rust seems a little messy, but I not sure what the best practices are for using futures cause that depends on `tokio` probably.

File: rust_snuba/rust_arroyo/src/processing/strategies/produce.rs (hunk 2 above)
Comment (dbanda) [reply rust_snuba/rust_arroyo/src/processing/strategies/produce.rs]: removed flush and made it place a future on the queue.

File: rust_snuba/rust_arroyo/src/processing/strategies/produce.rs (hunk 3 above)
Comment (dbanda) [reply rust_snuba/rust_arroyo/src/processing/strategies/produce.rs:88]: added a `panic` check to submit if closed is true

File: rust_snuba/rust_arroyo/src/processing/strategies/produce.rs (hunk 1 above)
Comment (dbanda) [reply rust_snuba/rust_arroyo/src/processing/strategies/produce.rs]: Ur right. removed

**Pull request merged**